server/application/rest/shelter.py

```diff
@@ -3,31 +3,67 @@
 for shelters in the server application.
 """
 import json
+
 from flask import Blueprint, Response, request
 from flask_cors import cross_origin
-from repository.mongo.shelter import ShelterRepo
-from use_cases.shelters.add_shelter_use_case import shelter_add_use_case
-from use_cases.shelters.list_shelters_use_case import shelter_list_use_case
+
+from authentication.authenticate_user import get_user_from_token
+
 from application.rest.status_codes import HTTP_STATUS_CODES_MAPPING
+
+from domains.resources import Resources
 from domains.shelter.shelter import Shelter
-from serializers.shelter import ShelterJsonEncoder
+
+from repository.mongo.shelter import ShelterRepo
+
 from responses import ResponseTypes
 
+from serializers.shelter import ShelterJsonEncoder
+
+from use_cases.authorization.is_authorized import is_authorized
+from use_cases.shelters.add_shelter_use_case import shelter_add_use_case
+from use_cases.shelters.list_shelters_use_case import shelter_list_use_case
+
 
 shelter_blueprint = Blueprint("shelter", __name__)
+
 @shelter_blueprint.route("/shelter", methods=["GET", "POST"])
 @cross_origin()
 def shelter():
     """
     On GET: The function returns a list of all shelters in the system.
+        Required permissions: READ access to Resources.SHELTER or system admin
+    
     On POST: The function adds a shelter to the system.
+        Required permissions: WRITE access to Resources.SHELTER or system admin
     """
     repo = ShelterRepo()
 
-    # add user authentication and authorization logic here
+    # Authentication check
+    auth_token = request.headers.get("Authorization")
+    if not auth_token:
+        return Response(
+            json.dumps({"message": "Authentication required"}),
+            mimetype="application/json",
+            status=HTTP_STATUS_CODES_MAPPING[
+                ResponseTypes.UNAUTHORIZED])
+    # Get the user email from the token
+    user_email = get_user_from_token(auth_token)
+    if not user_email:
+        return Response(
+            json.dumps({"message": "Invalid authentication token"}),
+            mimetype="application/json",
+            status=HTTP_STATUS_CODES_MAPPING[ResponseTypes.UNAUTHORIZED]
+        )
 
     if request.method == "GET":
-        # process the GET request parameters
+        # Authorization check for viewing shelters
+        if not is_authorized(repo, user_email, Resources.SHELTER):
+            return Response(
+                json.dumps({"message": "Unauthorized to view shelters"}),
+                mimetype="application/json",
+                status=HTTP_STATUS_CODES_MAPPING[
+                    ResponseTypes.FORBIDDEN])
```

Comment:
```suggestion
```
We'll need to allow volunteers to view shelters, otherwise they will not know where to go.